### PR TITLE
Remove KVO of progress in favor of using the NSURLSession delegate APIs

### PR DESCRIPTION
**Pull request by coreyfloyd**
The impetus for this change was a bug related to:
https://github.com/AFNetworking/AFNetworking/issues/3380

Which points to an issue where AFURLSessionManager could get in a state where it does not inform the AFURLSessionManagerTaskDelegate to un-observe cancelled (or otherwise completed) tasks before they are deallocated.
While the above issue is specific to background tasks/sessions, we have observed it on sessions with default configurations as well.

Instead of debugging the KVO issues, I decided to replace the KVO code with in favor of the appropriate NSURLSession delegate methods.
Running all unit tests shows that this approach appears to be working.

The pull request that made this change (https://github.com/AFNetworking/AFNetworking/pull/3187) mentions that its goal was to remove the need for AFNetworking API consumers to use a KVO API for getting progress.
This makes sense, but I was unable to see any reason why KVO was used internally for getting progress information rather than the NSURLSession Delegate methods.
Because of this, It isn't clear to me what the historical reasons were/are for KVO being needed to satisfy the progress requirements of the AFNetworking API.

With any luck, KVO is no longer needed and we can eliminate this class of crashes by eliminating KVO.

### CODE REVIEW

File: AFNetworking/AFURLSessionManager.m [hunk 1]
```diff
@@ -1073,6 +1069,12 @@ - (void)URLSession:(NSURLSession *)session
             totalUnitCount = (int64_t) [contentLength longLongValue];
         }
     }
+    
+    AFURLSessionManagerTaskDelegate *delegate = [self delegateForTask:task];
+    
+    if (delegate) {
```
Comment (joewalsh) [inline AFNetworking/AFURLSessionManager.m:1075]: does this need an `&& [delegate respondsToSelector...]`?

File: AFNetworking/AFURLSessionManager.m [hunk 2]
```diff
@@ -1199,6 +1201,13 @@ - (void)URLSession:(NSURLSession *)session
  totalBytesWritten:(int64_t)totalBytesWritten
 totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite
 {
+    
+    AFURLSessionManagerTaskDelegate *delegate = [self delegateForTask:downloadTask];
+    
+    if (delegate) {
```
Comment (joewalsh) [inline AFNetworking/AFURLSessionManager.m:1207]: ^ same question

File: AFNetworking/AFURLSessionManager.m [hunk 3]
```diff
@@ -1209,6 +1218,13 @@ - (void)URLSession:(NSURLSession *)session
  didResumeAtOffset:(int64_t)fileOffset
 expectedTotalBytes:(int64_t)expectedTotalBytes
 {
+    
+    AFURLSessionManagerTaskDelegate *delegate = [self delegateForTask:downloadTask];
+    
+    if (delegate) {
```
Comment (joewalsh) [inline AFNetworking/AFURLSessionManager.m:1224]: ^ same question

**Comment by coreyfloyd** [comment]
@joewalsh I am using the existing pattern of not checking if it responds to the selector. So will probably propose that as a change in the main repo just for safety. If that is acceptable then I will add the checks there.

**Pull request merged**